lib/msal-common/src/app/module/AuthorizationCodeModule.ts

```diff
@@ -107,77 +115,151 @@ export class AuthorizationCodeModule extends AuthModule {
+                }
+            }
+
+            const { tokenEndpoint } = acquireTokenAuthority;
+            const tokenReqParams = new ServerTokenRequestParameters(
                 this.clientConfig.auth.clientId,
                 tokenRequest,
                 codeResponse,
+                this.getAccount(),
                 this.getRedirectUri(),
                 this.cryptoObj
             );
 
-            const acquiredTokenResponse = await this.networkClient.sendPostRequestAsync<ServerAuthorizationTokenResponse>(
-                tokenEndpoint,
-                {
-                    body: tokenReqParams.createRequestBody(),
-                    headers: tokenReqParams.createRequestHeaders()
+            return this.getTokenResponse(tokenEndpoint, tokenReqParams, tokenRequest, codeResponse);
+        } catch (e) {
+            this.cacheManager.resetTempCacheItems(codeResponse && codeResponse.userRequestState);
+            this.account = null;
+            throw e;
+        }
+    }
+
+    async renewToken(request: TokenRenewParameters): Promise<TokenResponse> {
+        try {
+            if (!request) {
+                throw ClientAuthError.createEmptyTokenRequestError();
+            }
+
+            const acquireTokenAuthority = (request && request.authority) ? AuthorityFactory.createInstance(request.authority, this.networkClient) : this.defaultAuthorityInstance;
+            if (!acquireTokenAuthority.discoveryComplete()) {
+                try {
+                    await acquireTokenAuthority.resolveEndpointsAsync();
+                } catch (e) {
+                    throw ClientAuthError.createEndpointDiscoveryIncompleteError(e);
                 }
-            );
+            }
 
-            validateServerAuthorizationTokenResponse(acquiredTokenResponse);
-            const responseHandler = new ResponseHandler(this.clientConfig.auth.clientId, this.cacheStorage, this.cacheManager, this.cryptoObj);
-            const tokenResponse = responseHandler.createTokenResponse(acquiredTokenResponse, tokenReqParams.state);
-            this.account = tokenResponse.account;
-            return tokenResponse;
+            const account = request.account || this.getAccount();
+            const requestScopes = new ScopeSet(request.scopes, this.clientConfig.auth.clientId, true);
+            if (requestScopes.isLoginScopeSet()) {
+                // Check for login if id token is being renewed
+                if (!account) {
+                    throw ClientAuthError.createUserLoginRequiredError();
+                }
+            }
+
+            const cachedTokenItem = this.getCachedTokens(requestScopes, acquireTokenAuthority.canonicalAuthority, request.resource, account && account.homeAccountIdentifier);
+            const expiration = Number(cachedTokenItem.value.expiresOnSec);
+            const offsetCurrentTime = TimeUtils.now() + this.clientConfig.systemOptions.tokenRenewalOffsetSeconds;
```

Comment:
Label everything as seconds to make it clear.

Reply:
done.